console-subscriber/tests/support/task.rs

```diff
@@ -0,0 +1,228 @@
+use std::{error, fmt};
+
+use super::MAIN_TASK_NAME;
+
+/// An actual task
+///
+/// This struct contains the values recorded from the console subscriber
+/// client and represents what is known about an actual task running on
+/// the test's runtime.
+#[derive(Clone, Debug)]
+pub(super) struct ActualTask {
+    pub(super) id: u64,
+    pub(super) name: Option<String>,
+    pub(super) wakes: u64,
+    pub(super) self_wakes: u64,
+}
+
+impl ActualTask {
+    pub(super) fn new(id: u64) -> Self {
+        Self {
+            id,
+            name: None,
+            wakes: 0,
+            self_wakes: 0,
+        }
+    }
+}
+
+/// An error in task validation.
+pub(super) struct TaskValidationFailure {
+    /// The expected task whose expectations were not met.
+    expected: ExpectedTask,
+    /// The actual task which failed the validation
+    actual: Option<ActualTask>,
+    /// A textual description of the validation failure
+    failure: String,
+}
+
+impl error::Error for TaskValidationFailure {}
+
+impl fmt::Display for TaskValidationFailure {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.failure)
+    }
+}
+
+impl fmt::Debug for TaskValidationFailure {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.actual {
+            Some(actual) => write!(
+                f,
+                "Task Validation Failed!\n  Expected Task: {expected:?}\n  Actual Task:   {actual:?}\n  Failure:       {failure}",
+                expected = self.expected, failure = self.failure),
+            None => write!(
+                f,
+                "Task Validation Failed!\n  Expected Task: {expected:?}\n  Failure:       {failure}",
+                expected = self.expected, failure = self.failure),
+        }
+    }
+}
+
+/// An expected task.
+///
+/// This struct contains the fields that an expected task will attempt to match
+/// actual tasks on, as well as the expectations that will be used to validate
+/// which the actual task is as expected.
+#[derive(Clone, Debug)]
+pub(crate) struct ExpectedTask {
+    match_name: Option<String>,
+
+    expect_present: Option<bool>,
+    expect_wakes: Option<u64>,
+    expect_self_wakes: Option<u64>,
+}
+
+impl Default for ExpectedTask {
+    fn default() -> Self {
+        Self {
+            match_name: None,
+            expect_present: None,
+            expect_wakes: None,
+            expect_self_wakes: None,
+        }
+    }
+}
+
+impl ExpectedTask {
+    /// Returns whether or not an actual task matches this expected task.
+    ///
+    /// All matching rules will be run, if they all succeed, then `true` will
+    /// be returned, otherwise `false`.
+    pub(super) fn matches_actual_task(&self, actual_task: &ActualTask) -> bool {
+        if let Some(match_name) = &self.match_name {
```

Comment:
am I correct that this is saying that if an `ExpectedTask`'s `match_name` is not `Some`, the `ExpectedTask` will basically never do anything of value? why is `match_name` optional, then? couldn't we just always have there be a `match_name`, and have a constructor that takes the name?

Reply:
My idea is that in the future we will probably (maybe) want to match tasks on things other than the name. For example, I want to match a task associated with a resource or async op. While it's not implemented yet, it seemed to make sense to follow the pattern that a task can match on different things from the outset.